bqskit/passes/search/generators/discrete.py

```diff
@@ -0,0 +1,206 @@
+"""This module implements the DiscreteLayerGenerator class."""
+from __future__ import annotations
+
+import logging
+from typing import Sequence
+
+from bqskit.compiler.passdata import PassData
+from bqskit.ir.circuit import Circuit
+from bqskit.ir.gate import Gate
+from bqskit.ir.gates.constant.cx import CNOTGate
+from bqskit.ir.gates.constant.h import HGate
+from bqskit.ir.gates.constant.t import TGate
+from bqskit.ir.gates.parameterized.pauliz import PauliZGate
+from bqskit.ir.operation import Operation
+from bqskit.passes.search.generator import LayerGenerator
+from bqskit.qis.state.state import StateVector
+from bqskit.qis.state.system import StateSystem
+from bqskit.qis.unitary.unitarymatrix import UnitaryMatrix
+from bqskit.utils.typing import is_sequence
+
+
+_logger = logging.getLogger(__name__)
+
+
+class DiscreteLayerGenerator(LayerGenerator):
+    """
+    The DiscreteLayerGenerator class.
+
+    Expands circuits using only discrete gates.
+    """
+
+    def __init__(
+        self,
+        gateset: Sequence[Gate] = [HGate(), TGate(), CNOTGate()],
+        double_headed: bool = False,
+    ) -> None:
+        """
+        Construct a DiscreteLayerGenerator.
+
+        Args:
+            gateset (Sequence[Gate]): A sequence of gates that can be used
+                in the output circuit. These must be non-parameterized gates.
+                (Default: [HGate, TGate, CNOTGate])
+
+            double_headed (bool): If True, successors will be generated by
+                both appending and prepending gates. (Default: False)
```

Comment:
The short answers are yes and yes.

It's necessary because it let's you diagonalize instead of invert - many unitaries are far easier to diagonalize. There are certainly cases where circuits with the same structure are introduced, but a lot of that get's caught with hashing and the basic pruning functions already included.